Use dynamicDowncast<T> more in RenderTheme

annevk opened the pull request:
#### b6d576d54824560544da2e793ca9625b54d1989b
<pre>
Use dynamicDowncast&lt;T&gt; more in RenderTheme
<a href="https://bugs.webkit.org/show_bug.cgi?id=265299">https://bugs.webkit.org/show_bug.cgi?id=265299</a>

Reviewed by Chris Dumez and Darin Adler.

A few instances of downcast&lt;T&gt; remain due to asserts or not being
preceded by an equivalent is&lt;T&gt; check.

Also use RefPtr where possible and use clearer variable names for
impacted methods.

This includes a change by cdumez to Ref and RefPtr that makes the
various downcast methods work with const.

* Source/WTF/wtf/Ref.h:
(WTF::checkedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/RefPtr.h:
(WTF::checkedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::isChecked const):
(WebCore::RenderTheme::isIndeterminate const):
(WebCore::RenderTheme::isEnabled const):
(WebCore::RenderTheme::isFocused const):
(WebCore::RenderTheme::isPresenting const):
(WebCore::RenderTheme::hasListButton const):
(WebCore::RenderTheme::hasListButtonPressed const):
(WebCore::RenderTheme::paintSliderTicks):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustTextFieldStyle const):
(WebCore::RenderThemeIOS::paintTextFieldDecorations):
(WebCore::RenderThemeIOS::adjustMenuListButtonStyle const):
(WebCore::RenderThemeIOS::paintSliderTrack):
(WebCore::RenderThemeIOS::paintProgressBar):
(WebCore::RenderThemeIOS::isSubmitStyleButton const):
(WebCore::RenderThemeIOS::paintAttachment):
(WebCore::RenderThemeIOS::paintMeter):
(WebCore::RenderThemeIOS::paintSliderTicks):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):

Canonical link: <a href="https://commits.webkit.org/271170@main">https://commits.webkit.org/271170@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de2f8a7c4769940f8c86016f3ea3a37d9d530f5e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27505 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6147 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28756 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29733 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25179 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/27976 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8067 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3542 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24970 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27770 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4935 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23616 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4297 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/4462 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24613 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30369 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23921 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25133 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25055 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30596 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26682 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/4479 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2618 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5947 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/34140 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4936 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7394 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3557 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4868 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -396,25 +396,23 @@ StyleAppearance RenderTheme::autoAppearanceForElement(RenderStyle& style, const
-        auto& input = downcast<HTMLInputElement>(element.get());
-
-        if (input.isTextButton() || input.isUploadButton())
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(*elementPtr)) {
```
`*elementPtr` -> `element`

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -446,12 +444,11 @@ StyleAppearance RenderTheme::autoAppearanceForElement(RenderStyle& style, const
     }
 
-    if (is<HTMLSelectElement>(element)) {
+    if (RefPtr select = dynamicDowncast<HTMLSelectElement>(*elementPtr)) {
```
`*elementPtr` -> `element`

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -483,8 +480,8 @@ StyleAppearance RenderTheme::autoAppearanceForElement(RenderStyle& style, const
 
-        if (is<SearchFieldResultsButtonElement>(element)) {
-            if (!downcast<SearchFieldResultsButtonElement>(element.get()).canAdjustStyleForAppearance())
+        if (RefPtr button = dynamicDowncast<SearchFieldResultsButtonElement>(*elementPtr)) {
```
`*elementPtr` -> `element`

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1366,9 +1365,11 @@ bool RenderTheme::isSpinUpButtonPartHovered(const RenderObject& renderer) const
+bool RenderTheme::isPresenting(const RenderObject& renderer) const
 {
-    return is<HTMLInputElement>(o.node()) && downcast<HTMLInputElement>(*o.node()).isPresentingAttachedView();
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(renderer.node()))
```
This is a little more concise:
```
RefPtr input = dynamicDowncast<HTMLInputElement>(renderer.node());
return input && input->isPresentingAttachedView();
```

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1383,20 +1384,18 @@ bool RenderTheme::isDefault(const RenderObject& o) const
-
-    const auto& input = downcast<HTMLInputElement>(*(renderer.generatingNode()));
-    return input.list();
+    if (RefPtr input = dynamicDowncast<HTMLInputElement>(*(renderer.generatingNode())))
```
Can be written on 2 lines like above.

also, you introduced a bug here. The previous code used to null check `renderer.generatingNode()`, now it no longer does.

I would use:
```
RefPtr input = dynamicDowncast<HTMLInputElement>(renderer.generatingNode());
return input && input->list();
```

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1383,20 +1384,18 @@ bool RenderTheme::isDefault(const RenderObject& o) const
 bool RenderTheme::hasListButtonPressed(const RenderObject& renderer) const
 {
-    const auto* input = downcast<HTMLInputElement>(renderer.generatingNode());
+    RefPtr input = dynamicDowncast<HTMLInputElement>(*(renderer.generatingNode()));
```
Same bug as above, you are now assuming `renderer.generatingNode()` is non-null, which likely isn't OK or `renderer.generatingNode()` would return a C++ reference, not a raw pointer.

Should be:
```
RefPtr input = dynamicDowncast<HTMLInputElement>(renderer.generatingNode());
return input && buttonElement->active();
```

cdumez commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -1383,20 +1384,18 @@ bool RenderTheme::isDefault(const RenderObject& o) const
         return false;
 
-    if (auto* buttonElement = input->dataListButtonElement())
+    if (RefPtr buttonElement = input->dataListButtonElement())
```
This is not strictly needed as we're calling a trivial getter on buttonElement.

cdumez commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -614,9 +612,9 @@ static void adjustInputElementButtonStyle(RenderStyle& style, const HTMLInputEle
 bool RenderThemeIOS::paintSliderTrack(const RenderObject& box, const PaintInfo& paintInfo, const IntRect& rect)
 {
-    if (!is<RenderSlider>(box))
+    auto* renderSlider = dynamicDowncast<RenderSlider>(&box);
```
`&box` -> `box` or you're doing an unnecessary null check.

cdumez commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -708,9 +706,9 @@ static void adjustInputElementButtonStyle(RenderStyle& style, const HTMLInputEle
 bool RenderThemeIOS::paintProgressBar(const RenderObject& renderer, const PaintInfo& paintInfo, const IntRect& rect)
 {
-    if (!is<RenderProgress>(renderer))
+    auto* renderProgress = dynamicDowncast<RenderProgress>(&renderer);
```
`&renderer` -> `renderer`

cdumez commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1615,11 +1612,11 @@ static void paintAttachmentBorder(GraphicsContext& context, Path& borderPath)
 bool RenderThemeIOS::paintMeter(const RenderObject& renderer, const PaintInfo& paintInfo, const IntRect& rect)
 {
-    if (!is<RenderMeter>(renderer))
+    auto* renderMeter = dynamicDowncast<RenderMeter>(&renderer);
```
`&renderer` -> `renderer`

cdumez commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1299,22 +1297,21 @@ static void paintAttachmentBorder(GraphicsContext& context, Path& borderPath)
-    const RenderAttachment& attachment = downcast<RenderAttachment>(renderer);
+    auto* attachment = dynamicDowncast<RenderAttachment>(renderer);
+    if (!attachment)
+        return true;
```
This used to return false and now it returns true?

cdumez reviewed (approved):
LGTM with fix.

darinadler reviewed (approved):
Yes, those `Ref.h` and `RefPtr.h` changes seem right.

annevk committed:
Use dynamicDowncast<T> more in RenderTheme
https://bugs.webkit.org/show_bug.cgi?id=265299

Reviewed by Chris Dumez and Darin Adler.

A few instances of downcast<T> remain due to asserts or not being
preceded by an equivalent is<T> check.

Also use RefPtr where possible and use clearer variable names for
impacted methods.

This includes a change by cdumez to Ref and RefPtr that makes the
various downcast methods work with const.

* Source/WTF/wtf/Ref.h:
(WTF::checkedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WTF/wtf/RefPtr.h:
(WTF::checkedDowncast):
(WTF::downcast):
(WTF::dynamicDowncast):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autoAppearanceForElement const):
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::isChecked const):
(WebCore::RenderTheme::isIndeterminate const):
(WebCore::RenderTheme::isEnabled const):
(WebCore::RenderTheme::isFocused const):
(WebCore::RenderTheme::isPresenting const):
(WebCore::RenderTheme::hasListButton const):
(WebCore::RenderTheme::hasListButtonPressed const):
(WebCore::RenderTheme::paintSliderTicks):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustTextFieldStyle const):
(WebCore::RenderThemeIOS::paintTextFieldDecorations):
(WebCore::RenderThemeIOS::adjustMenuListButtonStyle const):
(WebCore::RenderThemeIOS::paintSliderTrack):
(WebCore::RenderThemeIOS::paintProgressBar):
(WebCore::RenderThemeIOS::isSubmitStyleButton const):
(WebCore::RenderThemeIOS::paintAttachment):
(WebCore::RenderThemeIOS::paintMeter):
(WebCore::RenderThemeIOS::paintSliderTicks):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::paintAttachment):

Canonical link: https://commits.webkit.org/271170@main
